src/gradient.jl

```diff
@@ -28,6 +28,11 @@ end
 
 @inline gradient!(out, f::F, x::SArray) where {F} = vector_mode_gradient!(out, f, x)
 
+@inline function gradient!(out::ImmutableDiffResult, f, x::SArray)
+    dual = vector_mode_dual_eval(f, x)
+    out = DiffResult(value(dual), extract_gradient(dual, x))
+end
```

Comment:
Do you mean, for example:
```
function extract_gradient!(out::DiffResult, dual::Dual)
    out = DiffBase.value!(out, value(dual))
    out = DiffBase.gradient!(out, partials(dual))
    return out
end
```
That would be simpler for sure. I worried about performance by constructing the ImmutableDiffResult in pieces, but it's probably negligible.